src/eval.rs

```diff
@@ -117,23 +117,130 @@ pub enum ThunkState {
+/// recursive merging of records and overriding (see the [RFC
+/// overriding](https://github.com/tweag/nickel/pull/330))
+#[derive(Clone, Debug, PartialEq)]
+pub enum InnerThunkData {
+    Standard(Closure),
+    Reversible {
+        orig: Rc<Closure>,
+        cached: Option<Closure>,
+    },
+}
+
 impl ThunkData {
+    /// Create new standard thunk data.
     pub fn new(closure: Closure) -> Self {
         ThunkData {
-            closure,
+            inner: InnerThunkData::Standard(closure),
             state: ThunkState::Suspended,
         }
     }
+
+    /// Create new reversible thunk data.
+    pub fn new_rev(orig: Closure) -> Self {
+        ThunkData {
+            inner: InnerThunkData::Reversible {
+                orig: Rc::new(orig),
+                cached: None,
+            },
+            state: ThunkState::Suspended,
+        }
+    }
+
+    /// Return a reference to the closure currently cached.
+    pub fn closure(&self) -> &Closure {
+        match self.inner {
+            InnerThunkData::Standard(ref closure) => closure,
+            InnerThunkData::Reversible {
+                ref orig,
+                cached: None,
+            } => orig,
+            InnerThunkData::Reversible {
+                cached: Some(ref closure),
+                ..
+            } => closure,
+        }
+    }
+
+    /// Return a mutable reference to the closure currently cached.
+    pub fn closure_mut(&mut self) -> &mut Closure {
+        match self.inner {
+            InnerThunkData::Standard(ref mut closure) => closure,
+            InnerThunkData::Reversible {
+                ref orig,
+                ref mut cached,
+            } => {
+                if cached.is_none() {
+                    *cached = Some((**orig).clone());
+                }
+                cached.as_mut().unwrap()
+            }
+        }
+    }
+
+    /// Consume the data and return the cached closure.
+    pub fn into_closure(self) -> Closure {
+        match self.inner {
+            InnerThunkData::Standard(closure) => closure,
+            InnerThunkData::Reversible { orig, cached: None } => {
+                Rc::try_unwrap(orig).unwrap_or_else(|rc| (*rc).clone())
+            }
+            InnerThunkData::Reversible {
+                cached: Some(cached),
+                ..
+            } => cached,
+        }
+    }
+
+    /// Update the cached closure.
+    pub fn update(&mut self, new: Closure) {
+        match self.inner {
+            InnerThunkData::Standard(ref mut closure) => *closure = new,
+            InnerThunkData::Reversible { ref mut cached, .. } => *cached = Some(new),
+        }
+
+        self.state = ThunkState::Evaluated;
+    }
+
+    /// Create fresh unevaluated thunk data from `self`, restored to its original state before the
+    /// first update. For standard thunk data, the content is unchanged and the state is conserved: in
+    /// this case, `restore()` is the same as `clone()`.
+    pub fn restore(&self) -> Self {
```

Comment:
I'm curious, why not the following?
```suggestion
    pub fn restore(&mut self) -> () {
```

Reply:
Because we want to create a new, separated "cache entry" (thunk) when reverting. For example:

```
let r1 = {foo = bar + 1, bar | default = 1} in
r1.foo + (let r2 = r1 & {bar = 2} in r2.foo + r2.foo) + r1.foo
```

On the first line, we allocate `r1` with a  thunk for `foo` inside, that is forced on the first operand of `+` to `2`. Then, we allocate an `r2` where we need to revert the thunks of fields of `r1`. But there is no reason to "revert" the thunk that is currently referenced by `r1.foo`, as it will always evaluate to `2`. We can't even share the same thunk at all between `r1` and `r2`: otherwise any evaluation of `r1.foo` or `r2.foo` could poison the cache of the other one. Rather, `r2` should allocate a fresh thunk, with the original expression taken from `r1.foo`, so that it can have its own independent caching for `foo`. This is what `restore` provides. But maybe there's a better name?

This also the reason for `Rc`ing `orig`, as several thunks will potentially share the same original expression but with different, independent cached values (hence the cached value is currently owned, though your suggestion is still interesting just for the sake of making code simpler)

Reply:
I see it now, thanks.